dotnet/src/Connectors/Connectors.Onnx.UnitTests/BertOnnxTextEmbeddingGeneration.cs

```diff
@@ -0,0 +1,383 @@
+
+                IList<ReadOnlyMemory<float>> examplesResults = await service.GenerateEmbeddingsAsync(
+                    examples.Select(s => upper ? s.ToUpperInvariant() : s).ToList());
+
+                string[] sortedExamples = examples
+                    .Zip(examplesResults)
+                    .OrderByDescending(p => TensorPrimitives.CosineSimilarity(inputResults[0].Span, p.Second.Span))
+                    .Select(p => p.First)
+                    .ToArray();
+
+                Assert.Equal(
+                    new string[]
+                    {
+                        "What is an amphibian?",
+                        "A frog is an amphibian.",
+                        "Amphibians are four-limbed and ectothermic vertebrates of the class Amphibia.",
+                        "Frogs, toads, and salamanders are all examples.",
+                        "A frog is green.",
+                        "Cos'è un anfibio?",
+                        "They are four-limbed and ectothermic vertebrates.",
+                        "A dog is a mammal.",
+                        "A tree is green.",
+                        "It's not easy bein' green.",
+                        "A dog is a man's best friend.",
+                        "You ain't never had a friend like me.",
+                        "Rachel, Monica, Phoebe, Joey, Chandler, Ross",
+                    },
+                    sortedExamples);
+            }
+        }
+    }
+
+    [Fact]
+    public async Task ValidateServiceMayBeUsedConcurrently()
+    {
+        using BertOnnxTextEmbeddingGenerationService service = await GetBgeMicroV2ServiceAsync();
+
+        string input = "What is an amphibian?";
+        IList<ReadOnlyMemory<float>> inputResults = await service.GenerateEmbeddingsAsync([input]);
+
+        string[] examples =
+        [
+            "A frog is an amphibian.",
+            "It's not easy bein' green.",
+            "A dog is a man's best friend.",
+            "A tree is green.",
+            "A dog is a mammal.",
+            "Rachel, Monica, Phoebe, Joey, Chandler, Ross",
+            "What is an amphibian?",
+            "Frogs, toads, and salamanders are all examples.",
+            "Cos'è un anfibio?",
+            "You ain't never had a friend like me.",
+            "Amphibians are four-limbed and ectothermic vertebrates of the class Amphibia.",
+            "A frog is green.",
+            "They are four-limbed and ectothermic vertebrates.",
+        ];
+
+        for (int trial = 0; trial < 10; trial++)
+        {
+            IList<ReadOnlyMemory<float>> examplesResults =
+                (await Task.WhenAll(examples.Select(e => service.GenerateEmbeddingsAsync([e])))).SelectMany(e => e).ToList();
+
+            string[] sortedExamples = examples
+                .Zip(examplesResults)
+                .OrderByDescending(p => TensorPrimitives.CosineSimilarity(inputResults[0].Span, p.Second.Span))
```

Comment:
It's out of scope for this PR, but do you think there's an argument for introducing a "find closest N" helper that avoids the need for `OrderByDescending` to sort all items (and only sort the top N) like in the SmartComponents.LocalEmbeddings package? This would also avoid consuming `O(num candidates)` memory on each search.

I know the counterargument could be "if you have so many candidates that it matters, then use an external vector DB" but for 1000-100000 (say) candidates, linear in-memory searches are fast enough for many use cases, but even then it's still desirable to maximize that speed.

Reply:
That's the purview of the IMemoryStore abstraction:
https://github.com/microsoft/semantic-kernel/blob/88bdb1155e148c98a72b171400cdc0ad0687ddae/dotnet/src/SemanticKernel.Abstractions/Memory/IMemoryStore.cs#L114
There are 10-15 implementations of it in this repo, including several four in-memory implementations. The SemanticTextMemory type then serves to marry one of these IMemoryStores with one of these ITextEmbeddingGenerators.

> TBH I wonder why we don't have a "TakeTop" or "TakeOrdered" (or whatever it would be called)

It doesn't actually require a custom public API. Order{By}{Descending} followed by Take is already optimized in the implementation to avoid a full sort of possible (lots of sequences of LINQ operators get special-cased). If you just take 1, it does the equivalent of a Min/Max{By}. If you take multiple, it does the equivalent of a partial quick sort (eg after partitioning, it only needs to recur into a side if the number being taken requires it). It could be optimized further for small numbers of items being taken via a priority queue, which is what https://github.com/dotnet/runtime/issues/96277 tracks... the trick is finding the right heiristic for when it's appropriate to cut over.